File: core/src/com/blakephillips/engine/ecs/systems/ai/StateSystem.java
package com.blakephillips.engine.ecs.systems.ai;

import com.badlogic.ashley.core.*;
import com.badlogic.ashley.systems.IteratingSystem;
import com.badlogic.ashley.utils.ImmutableArray;
import com.blakephillips.engine.ai.State;
import com.blakephillips.engine.ecs.components.ai.StateComponent;

public class StateSystem extends IteratingSystem {

    ComponentMapper<StateComponent> stateComponents = ComponentMapper.getFor(StateComponent.class);

    public StateSystem() {
        super(Family.all(StateComponent.class).get());
    }

    @Override
    protected void processEntity(Entity entity, float deltaTime) {
        StateComponent stateComponent = stateComponents.get(entity);

        if (stateComponent.state.stateStatus == State.StateStatus.IDLE) {
            stateComponent.state.enter();
        }
        if (stateComponent.state.stateStatus == State.StateStatus.RUNNING) {
            stateComponent.state.update(deltaTime);
        }
        // for now, just remove the state completely when exited.
        else if (stateComponent.state.stateStatus == State.StateStatus.EXITED) {
            entity.remove(StateComponent.class);
        }
    }
}

File: core/src/com/blakephillips/engine/ecs/systems/ai/StateSystem.kt
package com.blakephillips.engine.ecs.systems.ai

import com.badlogic.ashley.core.ComponentMapper
import com.badlogic.ashley.core.Entity
import com.badlogic.ashley.core.Family
import com.badlogic.ashley.systems.IteratingSystem
import com.blakephillips.engine.ai.State.StateStatus
import com.blakephillips.engine.ecs.components.ai.StateComponent

class StateSystem : IteratingSystem(Family.all(StateComponent::class.java).get()) {
    private var stateComponents: ComponentMapper<StateComponent> = ComponentMapper.getFor(StateComponent::class.java)
    override fun processEntity(entity: Entity, deltaTime: Float) {
        val stateComponent = stateComponents.get(entity)

        when (stateComponent.state.stateStatus) {
            StateStatus.IDLE -> stateComponent.state.enter()
            StateStatus.RUNNING -> stateComponent.state.update(deltaTime)
            StateStatus.COMPLETE -> entity.remove(StateComponent::class.java)
            else -> {}
        }
    }
}